lib/client/conntest/ssh.go

```diff
@@ -49,30 +59,239 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	ctxWithTimeout, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
 
-	if err := s.cd.UpdateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+	if err := teleClient.SSH(ctxWithTimeout, []string{"whoami"}, false); err != nil {
+		return s.handleErrFromSSH(ctx, connectionDiagnosticID, req.SSHPrincipal, err, processStdout)
+	}
+
+	connDiag, err := s.clt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		types.ConnectionDiagnosticTrace_NODE_PRINCIPAL,
+		fmt.Sprintf("%q user exists in target node", req.SSHPrincipal),
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.clt.UpdateConnectionDiagnostic(ctx, connDiag); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s SSHConnectionTester) proxyAddrs() (webProxyAddr string, sshProxyAddr string, err error) {
+	proxies, err := s.clt.GetProxies()
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	if len(proxies) == 0 {
+		return "", "", trace.NotFound("cluster has no proxies")
+	}
+
+	parsedAddrs, err := client.ParseProxyHost(proxies[0].GetAddr())
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	return parsedAddrs.WebProxyAddr, parsedAddrs.SSHProxyAddr, nil
+}
+
+func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDiagnosticID string, sshPrincipal string, sshError error, processStdout *bytes.Buffer) (types.ConnectionDiagnostic, error) {
+	// Either the node doesn't exist or the user doesn't have access to it.
+	if trace.IsNotFound(sshError) {
+		connDiag, err := s.clt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			types.ConnectionDiagnosticTrace_RBAC_NODE,
+			"Node not found. Ensure the Node exists and your role allows you to access it.",
+			sshError,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	if trace.IsConnectionProblem(sshError) {
+		connDiag, err := s.clt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			types.ConnectionDiagnosticTrace_CONNECTIVITY,
+			"Failed to connect to the Node. Ensure teleport is running.",
+			sshError,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	processStdoutString := strings.TrimSpace(processStdout.String())
+
+	if strings.Contains(sshError.Error(), "Process exited with status 255") && strings.HasPrefix(processStdoutString, "Failed to launch: user: unknown user") {
```

Comment:
Is "process exited with status 255" check required? Wouldn't checking for "unknown user" be enough?

Reply:
I can match only the process' stdout :+1: